### ISSUE TITLE
Regression MSVC 1.4.0 to 1.4.1

### ISSUE THREAD

**Issue by nacho** [behdad/harfbuzz#426]
I seem to be getting a regression when building harfbuzz with visual studio. This is the compile error I get now:

hb-unicode.obj : error LNK2001: unresolved external symbol _hb_ucdn_get_unicode_funcs
release\Win32\harfbuzz.dll : fatal error LNK1120: 1 unresolved externals
NMAKE : fatal error U1077: '"C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\BIN\link.EXE"' : return code '0x460'

**Comment by nacho** [comment]
So the reason is that hb_ucdn_get_unicode_funcs is internal, but why does it get externally exposed with msvc? Adding @fanc999

**Comment by fanc999** [comment]
Hi Nacho,

This is due to HarfBuzz defaulting on UCDN for its default unicode implementation unless it is explicitly disabled, and the update for that was not complete in config.h.win32.in.

I am working on a patch to the NMake Makefiles to fix this and to add a NO_UCDN option to allow UCDN to be disabled, and to fix introspection builds as well.

With blessings, and cheers!

**Comment by fanc999** [comment]
Hi @Nacho,

Can you try my master.msvc[1] branch and see whether things work for you, before I do a pull request?

With blessings, thank you!
[1]: https://github.com/fanc999/harfbuzz/tree/master.msvc

**Comment by nacho** [comment]
@fanc999 one thing I do not see in your patch is where you add the export to the method...

**Comment by fanc999** [comment]
Hi @Nacho,

No, it is not supposed to be exported, but hb-unicode.obj expects it to be there during linking of harfbuzz-vsX.dll.  The code that defines that symbol is not included in the build as the UCDN sources are not built, as we previously excluded the UCDN sources when we enable Glib or ICU, as the UCDN Unicode implementation was previously used as a fallback when ICU or Glib is not enabled.  It was later changed by Behdad to be the default Unicode implementation to fix bugs [1].

So [2] was done to fix the situation, and to make UCDN enabled by default.

[1]: https://github.com/fanc999/harfbuzz/commit/53543f8b3e19456279c86f711feba42ce976668c
[2]: https://github.com/fanc999/harfbuzz/commit/1bb8a84f2f83c5c8abec7751a355cf8f7166cb2c

With blessings, and cheers!

**Comment by behdad** [comment]
What's this about?

**Comment by nacho** [comment]
@fanc999 I wonder if we should stop with nmake and instead try to port harfbuzz to meson...

**Comment by fanc999** [comment]
Hi @nacho,

I think we could, pending on @behdad's and @ebraminio's viewpoints on this.

The issue for MSVC builds is that we need to use fallbacks for some non-GNOME dependencies as pkg-config files may well not be available from those dependencies in their MSVC-supporting build systems, like what is now done (or in progress of being done) in the Meson build files of the GTK+ stack.

This issue should have been fixed by PR #430, sorry for forgetting to write back here for that.

With blessings, and cheers!

**Comment by behdad** [comment]
We have another issue open for meson build. Thanks.
